### PR TITLE
🌱 flavor-gen: bump CPI manifests and use embed instead of hardcoding

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

Use `helm template --namespace=kube-system` to generate the CPI manifests used in flavor-gen.
This allows to streamline the update progress of CPI.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: packaging/flavorgen/cloudprovider/cloud-controller-manager.go [hunk 1]
```diff
@@ -35,295 +35,63 @@ const (
...
-		TypeMeta: metav1.TypeMeta{
-			Kind:       "ServiceAccount",
-			APIVersion: corev1.SchemeGroupVersion.String(),
-		},
+// CloudControllerManagerManifests returns the yaml manifests generated via
+// `helm template` from https://github.com/kubernetes/cloud-provider-vsphere/tree/master/charts/vsphere-cpi
+func CloudControllerManagerManifests() (string, error) {
+	objectList := []string{}
+
+	imageTagRemoval := regexp.MustCompile(`(image: .*):.*`)
```
Comment (sbueringer) [inline packaging/flavorgen/cloudprovider/cloud-controller-manager.go:43]: Let's add a short godoc comment about what we're replacing

File: packaging/flavorgen/cloudprovider/cloud-controller-manager.go [hunk 2]
```diff
@@ -35,295 +35,63 @@ const (
...
+	withoutTag := imageTagRemoval.ReplaceAll([]byte(cpiManifests), []byte(`${1}:$`+env.CPIImageKubernetesVersionVar))
+
+	// ClusterResourceSet does not support `kind: List`, we have to adjust the manifests
+	// to remove those lists and directly add the items instead.
```
Comment (sbueringer) [inline packaging/flavorgen/cloudprovider/cloud-controller-manager.go:47]: Do you remember what error you were getting?

I wonder if it is as easy as adding List to the schema that CRS uses

File: packaging/flavorgen/cloudprovider/cloud-controller-manager.go [hunk 3]
```diff
@@ -35,295 +35,63 @@ const (
...
-	}
+	return "---\n" + strings.Join(objectList, "---\n"), nil
 }
 
 // CloudControllerManagerConfigMap returns a ConfigMap containing data for the cloud config file.
 func CloudControllerManagerConfigMap(cloudConfig string) *corev1.ConfigMap {
 	return &corev1.ConfigMap{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      "vsphere-cloud-config",
+			Name:      "cloud-config",
```
Comment (sbueringer) [inline packaging/flavorgen/cloudprovider/cloud-controller-manager.go:90]: I assume this means the name in our structs was out-of-sync with the chart, right?

File: packaging/flavorgen/flavors/crs/cpi.go [hunk 4]
```diff
@@ -45,11 +43,6 @@ func addCPILabels(o metav1.Object, cpiInfraLabelValue string) {
 // CreateCrsResourceObjectsCPI creates the api objects necessary for CSI to function.
 // Also appends the resources to the CRS.
 func CreateCrsResourceObjectsCPI(crs *addonsv1.ClusterResourceSet) []runtime.Object {
-	serviceAccount := cloudprovider.CloudControllerManagerServiceAccount()
-	addCPILabels(serviceAccount, "service-account")
-	serviceAccountSecret := newSecret(serviceAccount.Name, serviceAccount)
-	appendSecretToCrsResource(crs, serviceAccountSecret)
-
 	credentials := map[string]string{}
```
Comment (sbueringer) [inline packaging/flavorgen/flavors/crs/cpi.go:46]: I assume all the things we drop here are part of cpi.yaml? (or something which provides whatever we need instead)

File: packaging/flavorgen/flavors/env/envsubts_consts.go [hunk 5]
```diff
@@ -18,34 +18,35 @@ limitations under the License.
...
+	ClusterClassNameVar                 = "${CLUSTER_CLASS_NAME}"
+	ControlPlaneMachineCountVar         = "${CONTROL_PLANE_MACHINE_COUNT}"
+	DefaultCloudProviderImage           = "gcr.io/cloud-provider-vsphere/cpi/release/manager:v1.2.1"
+	DefaultClusterCIDR                  = "192.168.0.0/16"
+	DefaultDiskGiB                      = 25
+	DefaultMemoryMiB                    = 8192
+	DefaultNumCPUs                      = 2
+	KubernetesVersionVar                = "${KUBERNETES_VERSION}"
+	CPIImageKubernetesVersionVar        = "${CPI_IMAGE_K8S_VERSION}"
+	CPIImageKubernetesVersionVarEscaped = `\${CPI_IMAGE_K8S_VERSION}`
```
Comment (sbueringer) [inline packaging/flavorgen/flavors/env/envsubts_consts.go:31]:
```suggestion
```
I assume leftover (unused)

File: packaging/flavorgen/cloudprovider/cloud-controller-manager.go (hunk 3 above)
Comment (chrischdi) [reply packaging/flavorgen/cloudprovider/cloud-controller-manager.go:90]: Yes, or manually edited so it has the prefix.

File: packaging/flavorgen/flavors/crs/cpi.go (hunk 4 above)
Comment (chrischdi) [reply packaging/flavorgen/flavors/crs/cpi.go:46]: Yes.

File: packaging/flavorgen/cloudprovider/cloud-controller-manager.go (hunk 2 above)
Comment (chrischdi) [reply packaging/flavorgen/cloudprovider/cloud-controller-manager.go:47]:
```
E0122 17:09:29.089263   18954 clusterresourceset_controller.go:362] "failed to apply ClusterResourceSet resource" err="creating object /v1, Kind=List : no matches for kind \"\" in version \"v1\"" controller="clusterresourceset" controllerGroup="addons.cluster.x-k8s.io" controllerKind="ClusterResourceSet" ClusterResourceSet="test-cluster-resource-set-7xskb/clusterresourceset-hybis4" namespace="test-cluster-resource-set-7xskb" name="clusterresourceset-hybis4" reconcileID="030779b8-e73c-44e3-bde7-9948b82e6455" Cluster="test-cluster-resource-set-7xskb/cluster-sdk18p" Resource kind="ConfigMap" Resource name="test-configmap"
```

Comment (chrischdi) [reply packaging/flavorgen/cloudprovider/cloud-controller-manager.go:47]: > Keep in mind that the Kubernetes API does not have a kind named List.
> 
> kind: List is a client-side, internal implementation detail for processing collections that might be of different kinds of object. Avoid depending on kind: List in automation or other code.

Source: https://kubernetes.io/docs/reference/using-api/api-concepts/

Comment (sbueringer) [reply packaging/flavorgen/cloudprovider/cloud-controller-manager.go:48]: Ah perfect. Okay that tells us we shouldn't fix it :D

**Comment by sbueringer** [comment]
Thx!

/approve

/assign @zhanggbj @laozc @fabriziopandini

**Comment by sbueringer** [comment]
Ah, looks like the linter isn't happy yet

File: packaging/flavorgen/cloudprovider/cloud-controller-manager.go [hunk 6]
```diff
@@ -35,295 +35,64 @@ const (
...
-		TypeMeta: metav1.TypeMeta{
-			Kind:       "ServiceAccount",
-			APIVersion: corev1.SchemeGroupVersion.String(),
-		},
+// CloudControllerManagerManifests returns the yaml manifests generated via
+// `helm template` from https://github.com/kubernetes/cloud-provider-vsphere/tree/master/charts/vsphere-cpi
+func CloudControllerManagerManifests() (string, error) {
+	objectList := []string{}
+
+	// Replace the hardcoded image tag by a variable for clusterctl. This way we can dynamically configure the verison of CPI.
```
Comment (sbueringer) [inline packaging/flavorgen/cloudprovider/cloud-controller-manager.go:43]:
```suggestion
	// Replace the hardcoded image tag by a variable for clusterctl. This way we can dynamically configure the version of CPI.
```

I would have ignored this nit, but golangci-lint didn't ...

**Review by zhanggbj — APPROVED** [review]
Thanks for the efforts!
/lgtm

**Pull request merged**